arangod/Aql/UpsertModifier.cpp

```diff
@@ -362,11 +446,29 @@ size_t UpsertModifier::nrOfWritesIgnored() const { return nrOfErrors(); }
 size_t UpsertModifier::getBatchSize() const { return _batchSize; }
 
 bool UpsertModifier::hasResultOrException() const noexcept {
-  return resultState() == ModificationExecutorResultState::HaveResult;
+  return std::visit(overload{
+                        [](NoResult) { return false; },
+                        [](Waiting) { return false; },
+                        [](HaveResult) { return true; },
+                        [](std::exception_ptr const&) { return true; },
+                    },
+                    _results);
 }
 
 bool UpsertModifier::hasNeitherResultNorOperationPending() const noexcept {
-  return resultState() == ModificationExecutorResultState::NoResult;
+  return std::visit(overload{
```

Comment:
I think this is a public method that is currently called only from the ModificationExecutor in an assertion. To be technically correct, I think this method must acquired the `_resultMutex` as well.
As the method is currently only used in a single assertion, we could alternatively remove it altogether.